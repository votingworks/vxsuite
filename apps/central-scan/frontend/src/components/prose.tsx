import styled from 'styled-components';

interface Props {
  compact?: boolean;
  maxWidth?: boolean;
  textCenter?: boolean;
}

export const Prose = styled('div')<Props>`
  margin: ${({ textCenter }) => (textCenter ? 'auto' : undefined)};
  max-width: ${({ maxWidth = true }) => (maxWidth ? '66ch' : undefined)};
  text-align: ${({ textCenter }) => (textCenter ? 'center' : undefined)};
  line-height: 1.2;
  @media (min-width: 480px) {
    line-height: 1.3;
  }
  & h1 {
    margin: 2em 0 1em;
    line-height: 1.1;
    font-size: 1.5em;
  }
  & h2 {
    margin: 1.5em 0 0.75em;
    font-size: 1.25em;
  }
  & h3 {
    font-size: 1.17em;
  }
  & h4 {
    font-size: 1em;
  }
  & h5 {
    font-size: 0.9em;
  }
  & h3,
  & h4,
  & h5,
  & p,
  & hr {
    margin-top: ${({ compact }) => (compact ? '0' : '1em')};
    margin-bottom: ${({ compact }) => (compact ? '0' : '1em')};
  }
  & h1 + h2 {
    margin-top: -0.75em;
  }
  & h1 + p,
  & h2 + p {
    margin-top: -0.75em;
  }
  & h3 + p,
  & h4 + p,
  & h5 + p {
    margin-top: ${({ compact }) => (compact ? 0 : '-1em')};
  }
  & > :not(.ignore-prose):first-child {
    margin-top: 0;
  }
  & > :not(.ignore-prose):last-child {
    margin-bottom: 0;
  }
  & dl {
    margin: 1em 0;
  }
  & hr {
    border: 0;
    border-top: 0.1em solid #000000;
  }
`;
