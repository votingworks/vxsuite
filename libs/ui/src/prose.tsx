import styled from 'styled-components';
import { Theme } from './themes';

// Prose!
// Readable text content with typographic hierarchy using simple semantic html.
//
// - Use basic text block html tags inside the <Prose> block. (see basic tags
//   below) Use the <Text> components for adding additional styles to html.
// - This is not a layout component; include it within a layout component to
//   define the space around the Prose component.
// - First and last child HTML tags of Prose have no margins so that a parent
//   layout component may define the layout. The HTML class "ignore-prose" HTML
//   class is applied to tag.
// - Width is constrained to 1.5 alphabets (66ch) to improve readability.
// - Spacing between elements is defined such that headings are closer to the
//   text that follows them.
//
// For example useage, search codebase for "<Prose" (without closing bracket)

export interface ProseProps {
  theme?: Theme;
  compact?: boolean;
  maxWidth?: boolean;
  textCenter?: boolean;
  textRight?: boolean;
}

export const Prose = styled('div')<ProseProps>`
  margin: ${({ textCenter }) => (textCenter ? '0 auto' : undefined)};
  max-width: ${({ maxWidth = true }) => (maxWidth ? '66ch' : undefined)};
  text-align: ${({ textCenter, textRight }) =>
    (textCenter && 'center') || (textRight && 'right')};
  line-height: 1.2;
  color: ${({ theme }) => theme.color};
  font-size: ${({ theme }) => theme.fontSize};
  & h1 {
    margin: 2em 0 1em;
    line-height: 1.1;
    font-size: 1.5em;
  }
  & h2 {
    margin: 1.5em 0 0.75em;
    font-size: 1.25em;
  }
  & h3 {
    font-size: 1.17em;
  }
  & h4 {
    font-size: 1em;
  }
  & h5 {
    font-size: 0.9em;
  }
  & h3,
  & h4,
  & h5,
  & p,
  & ol,
  & ul,
  & hr {
    margin-top: ${({ compact }) => (compact ? '0' : '1em')};
    margin-bottom: ${({ compact }) => (compact ? '0' : '1em')};
  }
  & h1 + h2 {
    margin-top: -0.75em;
  }
  & h1,
  & h2 {
    & + p,
    & + ol,
    & + ul {
      margin-top: -0.75em;
    }
  }
  & h3,
  & h4,
  & h5 {
    & + p,
    & + ol,
    & + ul {
      margin-top: ${({ compact }) => (compact ? 0 : '-1em')};
    }
  }
  & > :not(.ignore-prose):first-child {
    margin-top: 0;
  }
  & > :not(.ignore-prose):last-child {
    margin-bottom: 0;
  }
  & dl {
    margin: 1em 0;
  }
  & hr {
    border: 0;
    border-top: 0.1em solid #666666;
  }
`;
