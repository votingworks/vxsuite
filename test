#!/usr/bin/env bun

import { readFile } from 'fs/promises';
import { join } from 'path';

const { $ } = Bun;
export {};

const workspacePackageJson = await readFile(
  join(__dirname, 'package.json'),
  'utf8'
);
const { workspaces } = JSON.parse(workspacePackageJson);
const WIDTH = 100;

const results: { pkgPath: string; status: 'success' | 'failure' }[] = [];
let totalPassed = 0;
let totalFailed = 0;

for (const pkgPath of workspaces) {
  const tests = box(
    (async function* () {
      yield `→ bun run test --cwd ${pkgPath}`;
      const bunTest = $`bun run test --cwd ${pkgPath}`.quiet();

      const bunTestResult = await bunTest;

      yield* bunTestResult.stderr.toString('utf-8').split('\n');
      yield `${bunTestResult.exitCode === 0 ? '✅' : '❌'} ${pkgPath}`;

      if (bunTestResult.exitCode === 0) {
        results.push({ pkgPath, status: 'success' });
      } else {
        results.push({ pkgPath, status: 'failure' });
      }
    })()
  );

  let passed = 0;
  let failed = 0;

  for await (const line of tests) {
    const passedMatch = line.match(/(\d+) pass/);
    const failedMatch = line.match(/(\d+) fail/);

    if (passedMatch) {
      passed = parseInt(passedMatch[1], 10);
    }

    if (failedMatch) {
      failed = parseInt(failedMatch[1], 10);
    }

    console.log(line);
  }

  totalPassed += passed;
  totalFailed += failed;
}

console.log(
  `\n${totalPassed + totalFailed} tests: ${totalPassed} passed, ${totalFailed} failed`
);

for (const { pkgPath, status } of results) {
  console.log(`${status === 'success' ? '✅' : '❌'} ${pkgPath}`);
}

async function* box(lines: AsyncIterable<string>): AsyncIterable<string> {
  yield* lines;
}
