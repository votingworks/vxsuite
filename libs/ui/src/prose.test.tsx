import React from 'react';
import { render } from '../test/react_testing_library';

import { Prose } from './prose';

const proseContent = (
  <React.Fragment>
    <h1>Heading 1</h1>
    <h2>Heading 2</h2>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam
      modi magnam neque. Molestias recusandae, officia maiores nam pariatur
      earum qui inventore minus enim adipisci nemo voluptate at harum?
    </p>
    <h1>Heading 1</h1>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam
      modi magnam neque. Molestias recusandae, officia maiores nam pariatur
      earum qui inventore minus enim adipisci nemo voluptate at harum?
    </p>
    <h2>Heading 2</h2>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam
      modi magnam neque. Molestias recusandae, officia maiores nam pariatur
      earum qui inventore minus enim adipisci nemo voluptate at harum?
    </p>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam
      modi magnam neque. Molestias recusandae, officia maiores nam pariatur
      earum qui inventore minus enim adipisci nemo voluptate at harum?
    </p>
    <h3>Heading 3</h3>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam
      modi magnam neque. Molestias recusandae, officia maiores nam pariatur
      earum qui inventore minus enim adipisci nemo voluptate at harum?
    </p>
  </React.Fragment>
);
describe('renders Prose', () => {
  test('with defaults', () => {
    const { container } = render(<Prose>{proseContent}</Prose>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('with with non-default options', () => {
    const { container } = render(
      <Prose compact textCenter maxWidth={false}>
        {proseContent}
      </Prose>,
      {
        vxTheme: { colorMode: 'contrastLow', sizeMode: 'touchMedium' },
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('with right-aligned text', () => {
    const { container } = render(<Prose textRight>{proseContent}</Prose>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
