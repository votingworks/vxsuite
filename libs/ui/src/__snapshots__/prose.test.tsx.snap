// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Prose with defaults 1`] = `
.c0 {
  max-width: 66ch;
  line-height: 1.2;
}

.c0 h1 {
  margin: 2em 0 1em;
  line-height: 1.1;
  font-size: 1.5em;
}

.c0 h2 {
  margin: 1.5em 0 0.75em;
  font-size: 1.25em;
}

.c0 h3 {
  font-size: 1.17em;
}

.c0 h4 {
  font-size: 1em;
}

.c0 h5 {
  font-size: 0.9em;
}

.c0 h3,
.c0 h4,
.c0 h5,
.c0 p,
.c0 ol,
.c0 ul,
.c0 hr {
  margin-top: 1em;
  margin-bottom: 1em;
}

.c0 h1 + h2 {
  margin-top: -0.75em;
}

.c0 h1 + p,
.c0 h2 + p,
.c0 h1 + ol,
.c0 h2 + ol,
.c0 h1 + ul,
.c0 h2 + ul {
  margin-top: -0.75em;
}

.c0 h3 + p,
.c0 h4 + p,
.c0 h5 + p,
.c0 h3 + ol,
.c0 h4 + ol,
.c0 h5 + ol,
.c0 h3 + ul,
.c0 h4 + ul,
.c0 h5 + ul {
  margin-top: -1em;
}

.c0 > :not(.ignore-prose):first-child {
  margin-top: 0;
}

.c0 > :not(.ignore-prose):last-child {
  margin-bottom: 0;
}

.c0 dl {
  margin: 1em 0;
}

.c0 hr {
  border: 0;
  border-top: 0.1em solid #666666;
}

<div
  class="c0"
>
  <h1>
    Heading 1
  </h1>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h1>
    Heading 1
  </h1>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h3>
    Heading 3
  </h3>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
</div>
`;

exports[`renders Prose with right-aligned text 1`] = `
.c0 {
  max-width: 66ch;
  text-align: right;
  line-height: 1.2;
}

.c0 h1 {
  margin: 2em 0 1em;
  line-height: 1.1;
  font-size: 1.5em;
}

.c0 h2 {
  margin: 1.5em 0 0.75em;
  font-size: 1.25em;
}

.c0 h3 {
  font-size: 1.17em;
}

.c0 h4 {
  font-size: 1em;
}

.c0 h5 {
  font-size: 0.9em;
}

.c0 h3,
.c0 h4,
.c0 h5,
.c0 p,
.c0 ol,
.c0 ul,
.c0 hr {
  margin-top: 1em;
  margin-bottom: 1em;
}

.c0 h1 + h2 {
  margin-top: -0.75em;
}

.c0 h1 + p,
.c0 h2 + p,
.c0 h1 + ol,
.c0 h2 + ol,
.c0 h1 + ul,
.c0 h2 + ul {
  margin-top: -0.75em;
}

.c0 h3 + p,
.c0 h4 + p,
.c0 h5 + p,
.c0 h3 + ol,
.c0 h4 + ol,
.c0 h5 + ol,
.c0 h3 + ul,
.c0 h4 + ul,
.c0 h5 + ul {
  margin-top: -1em;
}

.c0 > :not(.ignore-prose):first-child {
  margin-top: 0;
}

.c0 > :not(.ignore-prose):last-child {
  margin-bottom: 0;
}

.c0 dl {
  margin: 1em 0;
}

.c0 hr {
  border: 0;
  border-top: 0.1em solid #666666;
}

<div
  class="c0"
>
  <h1>
    Heading 1
  </h1>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h1>
    Heading 1
  </h1>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h3>
    Heading 3
  </h3>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
</div>
`;

exports[`renders Prose with theme 1`] = `
.c0 {
  max-width: 66ch;
  line-height: 1.2;
  color: #666666;
  font-size: 10px;
}

.c0 h1 {
  margin: 2em 0 1em;
  line-height: 1.1;
  font-size: 1.5em;
}

.c0 h2 {
  margin: 1.5em 0 0.75em;
  font-size: 1.25em;
}

.c0 h3 {
  font-size: 1.17em;
}

.c0 h4 {
  font-size: 1em;
}

.c0 h5 {
  font-size: 0.9em;
}

.c0 h3,
.c0 h4,
.c0 h5,
.c0 p,
.c0 ol,
.c0 ul,
.c0 hr {
  margin-top: 1em;
  margin-bottom: 1em;
}

.c0 h1 + h2 {
  margin-top: -0.75em;
}

.c0 h1 + p,
.c0 h2 + p,
.c0 h1 + ol,
.c0 h2 + ol,
.c0 h1 + ul,
.c0 h2 + ul {
  margin-top: -0.75em;
}

.c0 h3 + p,
.c0 h4 + p,
.c0 h5 + p,
.c0 h3 + ol,
.c0 h4 + ol,
.c0 h5 + ol,
.c0 h3 + ul,
.c0 h4 + ul,
.c0 h5 + ul {
  margin-top: -1em;
}

.c0 > :not(.ignore-prose):first-child {
  margin-top: 0;
}

.c0 > :not(.ignore-prose):last-child {
  margin-bottom: 0;
}

.c0 dl {
  margin: 1em 0;
}

.c0 hr {
  border: 0;
  border-top: 0.1em solid #666666;
}

<div
  class="c0"
>
  <h1>
    Heading 1
  </h1>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h1>
    Heading 1
  </h1>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h3>
    Heading 3
  </h3>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
</div>
`;

exports[`renders Prose with with non-default options 1`] = `
.c0 {
  margin: 0 auto;
  text-align: center;
  line-height: 1.2;
}

.c0 h1 {
  margin: 2em 0 1em;
  line-height: 1.1;
  font-size: 1.5em;
}

.c0 h2 {
  margin: 1.5em 0 0.75em;
  font-size: 1.25em;
}

.c0 h3 {
  font-size: 1.17em;
}

.c0 h4 {
  font-size: 1em;
}

.c0 h5 {
  font-size: 0.9em;
}

.c0 h3,
.c0 h4,
.c0 h5,
.c0 p,
.c0 ol,
.c0 ul,
.c0 hr {
  margin-top: 0;
  margin-bottom: 0;
}

.c0 h1 + h2 {
  margin-top: -0.75em;
}

.c0 h1 + p,
.c0 h2 + p,
.c0 h1 + ol,
.c0 h2 + ol,
.c0 h1 + ul,
.c0 h2 + ul {
  margin-top: -0.75em;
}

.c0 h3 + p,
.c0 h4 + p,
.c0 h5 + p,
.c0 h3 + ol,
.c0 h4 + ol,
.c0 h5 + ol,
.c0 h3 + ul,
.c0 h4 + ul,
.c0 h5 + ul {
  margin-top: 0;
}

.c0 > :not(.ignore-prose):first-child {
  margin-top: 0;
}

.c0 > :not(.ignore-prose):last-child {
  margin-bottom: 0;
}

.c0 dl {
  margin: 1em 0;
}

.c0 hr {
  border: 0;
  border-top: 0.1em solid #666666;
}

<div
  class="c0"
>
  <h1>
    Heading 1
  </h1>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h1>
    Heading 1
  </h1>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h2>
    Heading 2
  </h2>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
  <h3>
    Heading 3
  </h3>
  <p>
    Lorem ipsum dolor sit amet consectetur adipisicing elit. Dolor omnis quam modi magnam neque. Molestias recusandae, officia maiores nam pariatur earum qui inventore minus enim adipisci nemo voluptate at harum?
  </p>
</div>
`;
